lib/RestructureCFG/InlineDivergentScopesPass.cpp

```diff
@@ -316,42 +385,182 @@ static void performIDS(DivergenceDescriptor &DivergenceDescriptor,
+    MultipleDivergences.erase(MultipleDivergences.begin() + LastElementIndex);
   }
 
+  // Decide if we need a new `Head` that collects the entry of the divergent
+  // scope. This is needed if one `Divergence` is composed by multiple
+  // `Successor`s, since in that case there is no clear entry to the inlined
+  // divergent scope. We therefore add a `Head` block that collects the entry
+  // point, so that this new node will dominate the eventual postdominator
+  // common to the multiple `Successor`s.
+  for (auto Divergence : MultipleDivergences) {
+    if (Divergence.DivergentSuccessors.size() > 1) {
+      createHead(Conditional, Divergence, PlaceHolderTarget);
+    }
+  }
+
+  // Move the `NonDivergentSuccessors` outgoing edges from `Conditional` so that
+  // they point to `Tail`
+  replaceSuccessors(ConditionalTerminator, NonDivergentSuccessors, Tail);
+
   // We remove from the `Terminator` of `Tail`, all the edges that
   // target `DivergentSuccessor`, since it will be only reached by
   // `Conditional`. We do this in two steps, we first substitute the original
   // successor with `PlaceHolderTarget`, and then we invoke
   // `simplifyTerminator`, which will take care of simplifying away the
   // unnecessary successors, both for `brcond`s and `switch`es.
-  for (BasicBlock *Successor : DivergentSuccessors) {
-    TailTerminator->replaceSuccessorWith(Successor, PlaceHolderTarget);
-  }
+  replaceSuccessors(TailTerminator, DivergentSuccessors, PlaceHolderTarget);
   simplifyTerminator(Tail, PlaceHolderTarget);
 
   // We add a `scope_closer` edge between the divergent exit node and
-  // the `Tail` node
-  ScopeGraphBuilder SGBuilder(F);
-  SGBuilder.addScopeCloser(Exit, Tail);
+  // the `Tail` node for all the exits
+  for (auto Divergence : MultipleDivergences) {
+    SGBuilder.addScopeCloser(Divergence.Exit, Tail);
+  }
+}
+
+/// Helper function which returns if a `DivergenceDescriptor` is compatible with
+/// all the ones already present in a `Collection`. Being compatible means that
+/// they insist on the same conditional, have non-overlapping `Exit` blocks, and
+/// do not have any overlapping `DivergentSuccessors`
+static bool isCompatible(const SmallVector<DivergenceDescriptor> &Collection,
+                         const DivergenceDescriptor &NewDescriptor) {
+  // Skip checking if collection is empty
+  if (Collection.empty())
+    return true;
+
+  // Check that Conditional matches
+  if (Collection[0].Conditional != NewDescriptor.Conditional)
+    return false;
+
+  // Check for non-overlapping Exit blocks
+  for (const auto &Existing : Collection) {
+    if (Existing.Exit == NewDescriptor.Exit)
+      return false;
+
+    // Check for non-overlapping DivergentSuccessors
+    for (BasicBlock *ExistingSucc : Existing.DivergentSuccessors) {
+      for (BasicBlock *NewSucc : NewDescriptor.DivergentSuccessors) {
+        if (ExistingSucc == NewSucc)
+          return false;
+      }
+    }
+  }
+
+  return true;
+}
```

Comment:
This should always be true by construction, as per our in-person discussion.

If anything, turn this into some logic for assertions.

Reply:
I improved the comments, and added a comment to the function specifying that this is used only inside an `assert` function https://github.com/revng/revng/blob/1b4cabba417d846b3f4e47ca33b3f6b7b0a15234/lib/RestructureCFG/InlineDivergentScopesPass.cpp#L454